ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixClusterManager.java

```diff
@@ -0,0 +1,563 @@
+
+  /**
+   * Return the unique {@link AmbryReplica} for a {@link AmbryPartition} on a {@link AmbryDataNode}.
+   * @param hostname the hostname associated with the {@link AmbryDataNode}.
+   * @param port the port associated with the {@link AmbryDataNode}.
+   * @param partitionString the partition id string associated with the {@link AmbryPartition}.
+   * @return the {@link AmbryReplica} associated with the given parameters.
+   */
+  AmbryReplica getReplicaForPartitionOnNode(String hostname, int port, String partitionString) {
+    for (AmbryReplica replica : dataNodeIdToReplicaIds.get(getDataNodeId(hostname, port))) {
+      if (replica.getPartitionId().toString().equals(partitionString)) {
+        return replica;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * An instance of this object is used to register as listner for Helix related changes in each datacenter.
+   */
+  private class ClusterChangeListener extends RoutingTableProvider implements LiveInstanceChangeListener {
+    final Set<String> allInstances = new HashSet<>();
+    private boolean liveInstanceChangeTriggered = false;
+    private boolean externalViewChangeTriggered = false;
+    private boolean configChangeTriggered = false;
+    private CountDownLatch initialized = new CountDownLatch(3);
+
+    @Override
+    public void onLiveInstanceChange(List<LiveInstance> liveInstances, NotificationContext changeContext) {
+      logger.info("Live instance change triggered with: " + liveInstances);
+      Set<String> liveInstancesSet = new HashSet<>();
+      for (LiveInstance liveInstance : liveInstances) {
+        liveInstancesSet.add(liveInstance.getInstanceName());
+        instanceNameToDataNodeIdMap.get(liveInstance.getInstanceName()).setState(HardwareState.AVAILABLE);
+      }
+      Set<String> downInstancesSet = Sets.difference(allInstances, liveInstancesSet);
+      for (String instanceName : downInstancesSet) {
+        instanceNameToDataNodeIdMap.get(instanceName).setState(HardwareState.UNAVAILABLE);
+      }
+      if (!liveInstanceChangeTriggered) {
+        liveInstanceChangeTriggered = true;
+        initialized.countDown();
+      }
+      helixClusterManagerMetrics.liveInstanceChangeTriggerCount.inc();
+    }
+
+    @Override
+    public void onExternalViewChange(List<ExternalView> externalViewList, NotificationContext changeContext) {
+      logger.info("ExternalView change triggered with: " + externalViewList);
+      super.onExternalViewChange(externalViewList, changeContext);
+
+      // No action taken at this time.
+
+      if (!externalViewChangeTriggered) {
+        externalViewChangeTriggered = true;
+        initialized.countDown();
+      }
+      helixClusterManagerMetrics.externalViewChangeTriggerCount.inc();
+    }
+
+    @Override
+    public void onConfigChange(List<InstanceConfig> configs, NotificationContext changeContext) {
+      logger.info("Config change triggered with: ", configs);
+      super.onConfigChange(configs, changeContext);
```

Comment:
Super class did not implement a `LiveInstanceChangeListener` interestingly. Anyway, based on @xiahome's comment, we now implement these listeners without extending the `RoutingTableProvider`.